preprocessing/science/refine_srf.py

```diff
@@ -0,0 +1,174 @@
+            myproj = pyproj.Proj(proj_string)
+            self.lon, self.lat = pyproj.transform(myproj, lla, self.x, self.y)
+        else:
+            self.lon, self.lat = self.x, self.y
+
+    def compute_time(self):
+        self.myt = np.linspace(0, (self.ndt - 1) * self.dt, self.ndt)
+
+    def write_SRF(self, fname):
+        with open(fname, "w") as fout:
+            fout.write("1.0\n")
+            fout.write("POINTS %d\n" % (self.nx * self.ny))
+            for j in range(self.ny):
+                for i in range(self.nx):
+                    fout.write("%g %g %g %g %g %e %g %g\n" % (self.lon[j, i], self.lat[j, i], self.depth[j, i], self.strike[j, i], self.dip[j, i], self.PSarea, self.t0[j, i], self.dt))
+                    fout.write("%g %g %d %f %d %f %d\n" % (self.rake[j, i], self.slip1[j, i], self.ndt, 0.0, 0, 0.0, 0))
+                    np.savetxt(fout, self.aSR[j, i, :], fmt="%g", newline=" ")
+                    fout.write("\n")
+        print("done writing", fname)
+
+
+def ReadSRF(fname):
+    with open(fname) as fid:
+        # version
+        line = fid.readline()
+        version = float(line)
+        if abs(version - 1.0) > 1e-03:
+            print("SRF version: %s not supported" % (line))
+            raise
+        line_el = fid.readline().split()
+        if line_el[0] != "PLANE":
+            print("no plane specified")
+            raise
+        line_el = fid.readline().split()
+        nx = int(line_el[2])
+        ny = int(line_el[3])
+        line_el = fid.readline().split()
+        if line_el[0] != "POINTS":
+            print("only one plane supported")
+            raise
+        # check that the plane data are consistent with the number of points
+        assert int(line_el[1]) == nx * ny
+        p1 = FaultPlane(nx, ny)
+        for j in range(ny):
+            for i in range(nx):
+                # first header line
+                line = fid.readline()
+                p1.lon[j, i], p1.lat[j, i], p1.depth[j, i], p1.strike[j, i], p1.dip[j, i], p1.PSarea, p1.t0[j, i], p1.dt = [float(v) for v in line.split()]
+                # second header line
+                line = fid.readline()
+                p1.rake[j, i], p1.slip1[j, i], ndt1, slip2, ndt2, slip3, ndt3 = [float(v) for v in line.split()]
+                ndt1 = int(ndt1)
+                if max(i, j) == 0:
+                    p1.ndt = ndt1
+                    p1.init_aSR()
+                lSTF = []
+                if ndt1 == 0:
+                    continue
+                if ndt1 > p1.ndt:
+                    print("this script assumes that ndt is not larger than the first ndt in the file", ndt1, p1.ndt)
+                    raise
+                while True:
+                    line = fid.readline()
+                    lSTF.extend(line.split())
+                    if len(lSTF) == ndt1:
+                        p1.aSR[j, i, 0:ndt1] = np.array([float(v) for v in lSTF])
```

Comment:
Here we just read the slip rate in u_1 direction, right? What about the slip rate in u_2 and u_3 direction?

Reply:
`Although	the	general	format	given	here	allows	for	slip	in	all	three	directions	(u1, u2, u3),	
most	cases	will	have	only	one	(u1)	or	two	(u1,	u2)	non-zero	slip	components.`
I will add a warning.